core/src/main/scala/com/pingcap/tispark/utils/ReflectionUtil.scala

```diff
@@ -43,85 +41,89 @@ object ReflectionUtil {
   //      isOrderSensitive: Boolean = false): RDD[U]
   //
   // Hereby we use reflection to support different Spark versions.
-  private val mapPartitionsWithIndexInternal: Method = spark_version match {
-    case "2.3.0" | "2.3.1" =>
-      tryLoadMethod(
-        "mapPartitionsWithIndexInternal",
-        mapPartitionsWithIndexInternalV1,
-        mapPartitionsWithIndexInternalV2
-      )
-    case _ =>
-      // Spark version >= 2.3.2
-      tryLoadMethod(
-        "mapPartitionsWithIndexInternal",
-        mapPartitionsWithIndexInternalV2,
-        mapPartitionsWithIndexInternalV1
-      )
-  }
-
-  // Spark HDP Release may not compatible with official Release
-  // see https://github.com/pingcap/tispark/issues/1006
-  private def tryLoadMethod(name: String, f1: () => Method, f2: () => Method): Method = {
-    try {
-      f1.apply()
-    } catch {
-      case _: Throwable =>
-        try {
-          f2.apply()
-        } catch {
-          case _: Throwable =>
-            throw ScalaReflectionException(
-              s"Cannot find reflection of Method $name, current Spark version is %s"
-                .format(spark_version)
-            )
-        }
-    }
+  case class ReflectionMapPartitionWithIndexInternal(
+    rdd: RDD[InternalRow],
+    internalRowToUnsafeRowWithIndex: (Int, Iterator[InternalRow]) => Iterator[UnsafeRow]
+  ) {
+    // Spark HDP Release may not compatible with official Release
+    // see https://github.com/pingcap/tispark/issues/1006
+    def invoke(): RDD[InternalRow] =
+      spark_version match {
+        case "2.3.0" | "2.3.1" =>
+          try {
+            mapPartitionsWithIndexInternalV1(rdd, internalRowToUnsafeRowWithIndex)
+          } catch {
+            case _: Throwable =>
```

Comment:
We catch all exception here. What if the exception thrown by spark is not what we want?

Reply:
nice! fixed